### PR TITLE
Initial work to support different Lateral join Types

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java [hunk 1]
```diff
@@ -36,28 +36,22 @@
...
+    private final JoinType type;
 
     @JsonCreator
     public LateralJoinNode(
             @JsonProperty("id") PlanNodeId id,
             @JsonProperty("input") PlanNode input,
             @JsonProperty("subquery") PlanNode subquery,
             @JsonProperty("correlation") List<Symbol> correlation,
-            @JsonProperty("type") Type type)
+            @JsonProperty("type") JoinType type)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java]: Are all values of `JoinType` valid for `LateralJoinNode`?
If not add a precondition check.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformUncorrelatedLateralToJoin.java [hunk 2]
```diff
@@ -44,6 +44,10 @@ public Pattern getPattern()
     {
         LateralJoinNode lateralJoinNode = (LateralJoinNode) node;
 
+        if (lateralJoinNode.getType() == JoinType.INNER) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformUncorrelatedLateralToJoin.java]: Did you mean `!=`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java]: They are valid as lateral join should behave like regular join. See discussion here: https://github.com/prestodb/presto/issues/8360

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java [hunk 3]
```diff
@@ -44,16 +46,19 @@
...
             @JsonProperty("id") PlanNodeId id,
             @JsonProperty("input") PlanNode input,
             @JsonProperty("subquery") PlanNode subquery,
             @JsonProperty("correlation") List<Symbol> correlation,
-            @JsonProperty("type") JoinType type)
+            @JsonProperty("type") JoinType type,
+            @JsonProperty("filter") Optional<Expression> filter)
     {
         super(id);
+        this.filter = filter;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java]: checkArgument initially that it is empty with TODO?

**Pull request closed without merging**